src/libraries/Microsoft.Extensions.Hosting/src/Internal/ServiceFactoryAdapter.cs

```diff
@@ -27,7 +27,7 @@ public object CreateBuilder(IServiceCollection services)
         {
             if (_serviceProviderFactory == null)
             {
-                _serviceProviderFactory = _factoryResolver(_contextResolver());
+                _serviceProviderFactory = _factoryResolver!(_contextResolver!());
```

Comment:
I think this would be better modeled as a `Debug.Assert`. Either `_serviceProviderFactory` or `_factoryResolver` needs to be non-null.

Reply:
Compiler doesn't understand this

![image](https://user-images.githubusercontent.com/6609929/161640410-f808f807-652f-4022-9a8f-2e2728b71845.png)

Reply:
```C#
            if (_serviceProviderFactory == null)
            {
                Debug.Assert(_factoryResolver != null);
                _serviceProviderFactory = _factoryResolver(_contextResolver());
```

Reply:
Oh, you mean like this. What about `_contextResolver`? It should be added to assert too?
PS: It's a pity that there's no attribute to say "If this member is null, than this isn't"